### PR TITLE
LoopVersioner tuning changes

### PR DESCRIPTION
**Pull request by JamesKingdon**
When the aggressive loop versioner option is set (PR #1369) this
change prevents calls to detectInvariantTrees and
buildNullCheckComparisonsTree based on opt level.

Signed-off-by: JamesKingdon <jkingdon@ca.ibm.com>

### CODE REVIEW

File: compiler/optimizer/LoopVersioner.cpp [hunk 1]
```diff
@@ -4131,9 +4135,15 @@ void TR_LoopVersioner::versionNaturalLoop(TR_RegionStructure *whileLoop, List<TR
 
    // Construct the tests for invariant expressions that need
    // to be null checked.
-   //
-   if (comp()->cg()->performsChecksExplicitly() ||
-       (comp()->getMethodHotness() >= veryHot))
+   
+   // default hotness threshold
+   int hotnessThreshold = hot;
```
Comment (andrewcraik) [inline compiler/optimizer/LoopVersioner.cpp]: This should not be int - it should be the type of the hotness enum

File: compiler/optimizer/LoopVersioner.cpp [hunk 2]
```diff
@@ -527,9 +527,13 @@ int32_t TR_LoopVersioner::performWithoutDominators()
...
          {
-         if (((debug("nullCheckVersion") || comp()->cg()->performsChecksExplicitly() ||
-               (comp()->getMethodHotness() >= veryHot)) &&
-               nullChecksWillBeEliminated) ||
+         // default hotness threshold
+         int hotnessThreshold = hot;
+         if (TR::Options::getCmdLineOptions()->getOption(TR_EnableAggressiveLoopVersioning))
+            hotnessThreshold = maxHotness; // threshold which can't be matched by the > operator
+         
+         if (( (debug("nullCheckVersion") || comp()->cg()->performsChecksExplicitly() || (comp()->getMethodHotness() > hotnessThreshold)) &&
```
Comment (andrewcraik) [inline compiler/optimizer/LoopVersioner.cpp]: can we format this more nicely so it is easier to match with everything else?

File: compiler/optimizer/LoopVersioner.cpp [hunk 3]
```diff
@@ -4131,9 +4135,15 @@ void TR_LoopVersioner::versionNaturalLoop(TR_RegionStructure *whileLoop, List<TR
...
-   //
-   if (comp()->cg()->performsChecksExplicitly() ||
-       (comp()->getMethodHotness() >= veryHot))
+   
+   // default hotness threshold
+   int hotnessThreshold = hot;
+
+   // If aggressive loop versioning is requested, don't call buildNullCheckComparisonsTree based on hotness
+   if (TR::Options::getCmdLineOptions()->getOption(TR_EnableAggressiveLoopVersioning))
+     hotnessThreshold = maxHotness; // threshold which can't be matched by the > operator
```
Comment (andrewcraik) [inline compiler/optimizer/LoopVersioner.cpp]: Probably want to put some trace in here to chat about what is going on?

**Comment by JamesKingdon** [comment]
@andrewcraik 
Re-typed, re-formatted, added trace and ran a PB

**Comment by vijaysun-omr** [comment]
Looks okay to me. Maybe rebase and if checks pass, I can consider for merging.

**Comment by vijaysun-omr** [comment]
Checks have passed. Merging.

**Pull request merged**